add: prepared statements documentation (#107)
scylladb-zpp-2025-python-rs-driver/python-rs-driver: docs/source/topics/using/prepared-statements.md

# Prepared Statements

Prepared statements are parsed by ScyllaDB once and cached. Every subsequent execution only sends the bound parameter values (and prepared statement id, which is a small hash instead of the full statement string to yet be parsed), reducing network traffic and CPU usage. Once created, prepared statements should be reused with different bind variables. Prepared statements use `?` marker to denote bind variables in the literal statement.

You can prepare a `Statement` or a literal `str`. One reason for using `Statement` instead of literal `str` is to create a new instance with desired configuration options or [`ExecutionProfile`](execution-profiles.md) using `with_*` methods before `await`ing using `session.prepare()`:

```python
insert_prepared_statement = await session.prepare(
    "INSERT INTO users (id, name, age) VALUES (?, ?, ?)"
)

# Or you can use a `Statement` with configuration options or attach an `ExecutionProfile`

> 

Reviewer: Agreed. Let's show imports.

> insert_prepared_statement = await session.prepare(
>     "INSERT INTO users (id, name, age) VALUES (?, ?, ?)"
> )
> 
> # Or you can use a `Statement` with configuration options or attach an `ExecutionProfile`
> 
> configured_statement = Statement("INSERT INTO users (id, name, age) VALUES (?, ?, ?)").with_request_timeout(10.0)
> 
> profile = ExecutionProfile(timeout=10.0, consistency=Consistency.LocalQuorum)
> configured_statement = Statement("INSERT INTO users (id, name, age) VALUES (?, ?, ?)").with_execution_profile(profile)

Reviewer: ```suggestion
configured_statement = Statement("INSERT INTO users (id, name, age) VALUES (?, ?, ?)").with_request_timeout(10.0)

profile = ExecutionProfile(timeout=10.0, consistency=Consistency.LocalQuorum)
configured_statement = configured_statement.with_execution_profile(profile)
```

Author: This change is showing implicitly that setting a timeout in a profile and with `with_request_timeout()` is ok. It is correct but redundant as `with_request_timeout()` shadows the profile's option so it doesn't make sense to use both profile and fined grained option on the same object.

Reviewer: Can you add the above intention description as a comment?

configured_prepared = await session.prepare(configured_statement)

# All `with_*` methods use immutable objects, so they do not modify the original statement.

statement = Statement("INSERT INTO users (id, name, age) VALUES (?, ?, ?)")
new_statement = statement.with_execution_profile(profile)
assert statement.execution_profile is None
assert new_statement.execution_profile is not None

# The same `with_*` methods can be used with already prepared statements. You don't need to call `prepare` again.

statement = Statement("INSERT INTO users (id, name, age) VALUES (?, ?, ?)")
prepared = await session.prepare(statement)
new_statement = prepared.with_request_timeout(10.0).with_consistency(Consistency.LocalQuorum)
assert statement.request_timeout is Unset
assert new_statement.request_timeout == 10.0
assert statement.consistency is None

> assert new_statement.consistency is Consistency.LocalQuorum

Reviewer: Hmm. Can there be any problem with `is` comparison correctness?

> assert new_statement.consistency is Consistency.LocalQuorum

Reviewer: cc @Lorak-mmk @patrycja-ziemkiewicz

> assert new_statement.consistency is Consistency.LocalQuorum

Reviewer 2: I sure hope not, `is` is widely used so we need it to work with our enums.
Someone should check that.

> assert new_statement.consistency is Consistency.LocalQuorum

Author: `is` works as it's an `IntEnum`

# Prepared Statements

> Prepared statements are parsed by ScyllaDB once and cached. Every subsequent execution only sends the bound parameter values (and prepared statement id, which is a small hash instead of the full statement string to yet be parsed), reducing network traffic and CPU usage. Once created, prepared statements should be reused with different bind variables. Prepared statements use `?` marker to denote bind variables in the literal statement.

Reviewer: ```suggestion
Prepared statements are parsed by ScyllaDB once and cached. Every subsequent execution only sends the bound parameter values (and prepared statement id, which is a small hash instead of the full statement string to yet be parsed), reducing network traffic and CPU usage. Once created, prepared statements should be reused with (possibly) different bind variables. Prepared statements use `?` marker to denote bind variables in the literal statement.
```

Reviewer: This mentions only positional bound parameters. You could also mention named bound parameters.

You can prepare a `Statement` or a literal `str`. One reason for using `Statement` instead of literal `str` is to create a new instance with desired configuration options or [`ExecutionProfile`](execution-profiles.md) using `with_*` methods before `await`ing using `session.prepare()`:

```python
from scylla.enums import Consistency
from scylla.execution_profile import ExecutionProfile
from scylla.statement import Statement
from scylla.types import Unset

# NOTE: This snippet is intended to run inside an `async def` where `session` is a connected Session.

insert_prepared_statement = await session.prepare(
    "INSERT INTO users (id, name, age) VALUES (?, ?, ?)"
)

# Or you can use a `Statement` with configuration options or attach an `ExecutionProfile`

configured_statement = Statement("INSERT INTO users (id, name, age) VALUES (?, ?, ?)").with_request_timeout(10.0)
```

> Worth noting that you can set both `with_*` options and an `ExecutionProfile`, but driver still adheres to hierarchy described in [Execution Profiles](execution-profiles.md)

Reviewer: ```suggestion
You can set both `with_*` options and an `ExecutionProfile`, but driver still adheres to hierarchy described in [Execution Profiles](execution-profiles.md), meaning that settings on a statement take precedence over settings on an execution profile.
```

```python
profile = ExecutionProfile(timeout=10.0, consistency=Consistency.LocalQuorum)

> configured_statement = Statement("INSERT INTO users (id, name, age) VALUES (?, ?, ?)").with_execution_profile(profile)

Reviewer: ⛏️ Let's put `.with_execution_profile` in a new line.
